### ISSUE TITLE
Level 6 test 6_2 logic

### ISSUE THREAD

**Issue by Seibell** [nus-cs2030/2122-s2#367]
# Summary
Hi all, im having this issue that some other people on the forum are experiencing where:
![image](https://user-images.githubusercontent.com/66213395/163731585-955ee09a-8642-4fad-a14d-a5395695c480.png)

## Description 
For my current implementation: My server contains an ImList<Customer> along with the maxQueueLength.

For arrive: 

1. Server is available > `Serve Event` > if server has people in his queue, i will remove the person at 0 index + `Done Event`

2. Server queue size less than maxQueueLength > `Wait Event` + adds customers into server queue > `Serve Event` > `Done Event`

3. else `Leave Event`

By this logic, if there are multiple customers in the server queue. When the first person in the queue is served, the person is removed and the server should update as having the second person in the queue only. (who is to be served at the next available time for the server).

Is there perhaps some error in my logic? Or am i not updating my servers properly (currently they are updated at every event other than Arrive & Leave)

**Comment by WenJett** [comment]
Hello, 

From your description I believe your logic is fine. You should look at your Arrive Event 2nd pointer as your Serve Event is queued up for all the customers in queue with the server's next immediate free time, which is correct for the person at index 1, but not true for those at index 2 onwards (since the person at index 2 needs to wait for person at index 1 and 0 to resolve their service).  

In short, what I did was that serve event is only executed for the person at index 1, and had a pending wait event for those at index 2 onwards until they are at index 1, then execute the serve event for them.

**Comment by vveining** [comment]
i'm not sure what your implementation is like, but for me i had the same error so i can let you know what went wrong with mine.

server 1 is serving customer 7 until 10.484, so the next available time of server 1 is 10.484.

while customer 7 is being served:
9.105 9 arrives
9.105 9 waits at 1 
9.160 10 arrives
9.160 10 waits at 1

executing the events of `9 waits at 1` and `10 waits at 1` results in the same serve event time for both 9 and 10 by server 1, as the execute in wait utilises the same next avail time of 10.484

**Comment by ReganKvioke** [comment]
@vveining Then how did you overcome this problem? I believe I am having the same issue as you

**Comment by vveining** [comment]
> @vveining Then how did you overcome this problem? I believe I am having the same issue as you

like what some have suggested in the forum, i made a "pending" event. customers who wait will go through the sequence of: 
arrive -> wait -> pending

in `Pending.execute` i check if 
- the waiting customer is first in the waitlist queue
- the server that the customer is waiting for is free
if yes to both, then `Pending.execute` returns a new `Serve`, otherwise you return another `Pending` event except you update the pending time with the next free time of the server. when i return a new `Serve`, i remove the customer from the server's waitlist, so all the remaining waiting customers shift up the waitlist queue
